### PR TITLE
Add error if spawn placements are missing for entities that spawn in-world

### PR DESCRIPTION
**Pull request by ThatGravyBoat**
### Description

This PR adds an error in dev if there are any entities added to the `MobSpawnSettings` for biomes and the entity does not register a spawn placement. The reason for this is that if you do not register a spawn placement the entity can spawn anywhere which includes in the are and is usually a bug 99.9% of the time and can be hard to track down if a developer does not know where to look and does not know they need to register to such an event this logging will help with such an issue and help prevent such bug from reaching production. If a developer does need to not have their entities restricted they can register with NO_RESTRICTIONS as what vanilla does if none is registered is just default to `NO_RESTRICTIONS` and always true for the predicate.

### CODE REVIEW

File: patches/net/minecraft/world/entity/SpawnPlacements.java.patch [hunk 1]
```diff
@@ -8,13 +8,18 @@
...
          EntityType<T> p_217075_, ServerLevelAccessor p_217076_, MobSpawnType p_217077_, BlockPos p_217078_, RandomSource p_217079_
      ) {
          SpawnPlacements.Data spawnplacements$data = DATA_BY_TYPE.get(p_217075_);
 -        return spawnplacements$data == null || ((SpawnPredicate<T>)spawnplacements$data.predicate).test(p_217075_, p_217076_, p_217077_, p_217078_, p_217079_);
 +        boolean vanillaResult = spawnplacements$data == null || spawnplacements$data.predicate.test((EntityType)p_217075_, p_217076_, p_217077_, p_217078_, p_217079_);
 +        return net.neoforged.neoforge.event.EventHooks.checkSpawnPlacements(p_217075_, p_217076_, p_217077_, p_217078_, p_217079_, vanillaResult);
++    }
++
++    // neo: Added to allow for checking if an entity has a spawn placement
++    public static boolean hasPlacement(EntityType<?> type) {
```
Comment (sciwhiz12) [inline patches/net/minecraft/world/entity/SpawnPlacements.java.patch]: The `Neo:` should be capitalized.

**Review by sciwhiz12 — DISMISSED** [review]
Hello! @ThatGravyBoat, do you intend to update this PR to 1.21? If so, please address the change I requested.

**Comment by ThatGravyBoat** [comment]
> Hello! @ThatGravyBoat, do you intend to update this PR to 1.21? If so, please address the change I requested.

Ive updated it to 1.21

**Comment by TelepathicGrunt** [comment]
Question, could we also add an error message if a modder puts the mob into the wrong mob type in the biomes? Basically, if a mob is a Creature, but is put under Monster category in biome, the mob will spawn nonstop until the monster cap is reached by a different mob. 

Note, iirc, vanilla is bugged where they put ocelot in the wrong biome mob category so you may need to explicitly exempt that. Maybe only check modded mobs in vanilla biomes and check all mobs in modded biomes?

**Comment by ThatGravyBoat** [comment]
> Question, could we also add an error message if a modder puts the mob into the wrong mob type in the biomes? Basically, if a mob is a Creature, but is put under Monster category in biome, the mob will spawn nonstop until the monster cap is reached by a different mob.
> 
> Note, iirc, vanilla is bugged where they put ocelot in the wrong biome mob category so you may need to explicitly exempt that. Maybe only check modded mobs in vanilla biomes and check all mobs in modded biomes?

If we were to add the message for bad spawn categories wouldn't it be better to fix vanilla doing it wrong instead of checking certain cases? 
The other problem is unlike spawn placements you cant tell it that you want it to not contribute to the spawn cap. 
Maybe even instead of a message, neo changes how the loading of the json works and automatically put them into the correct categories, ignoring where they were actually put in the json? (Would be another PR ofc)

**Review by TelepathicGrunt — DISMISSED** [review]

**Review by TelepathicGrunt — DISMISSED** [review]

**Comment by TelepathicGrunt** [comment]
Tested locally with my mod
image

Seems to be working. will give 24 hours before merge unless someone has some objections or suggestions

**Review by TelepathicGrunt — APPROVED** [review]

**Pull request merged**